net/pfSense-pkg-Quagga_OSPF/files/usr/local/pkg/quagga_ospfd.inc

```diff
@@ -323,39 +322,82 @@ function quagga_ospfd_install_conf() {
 			fwrite($fd, $bgpdaddmd5file);
 			fclose($fd);
 		}
-
 		if ($bgpddelmd5file != "") {
 			$fd = fopen("{$quagga_config_base}/bgpddelmd5pw.conf", "w");
 			fwrite($fd, $bgpddelmd5file);
 			fclose($fd);
 		}
 	}
-        /* Make ospf6 config */
+	/* Make ospf6 config */
 	if (isset($config['installedpackages']['quaggaospfdraw']['config'][0]['ospf6d'])
 	    && !empty($config['installedpackages']['quaggaospfdraw']['config'][0]['ospf6d'])) {
 		// if there is a raw config specified in the config.xml use that instead of the assisted config
 		$ospf6dconffile = str_replace("\r","",base64_decode($config['installedpackages']['quaggaospfdraw']['config'][0]['ospf6d']));
 	} else {
-        	$ospf6dconffile = "";
+		$ospf6dconffile = "";
 	}
 
 	$fd = fopen("{$quagga_config_base}/ospf6d.conf", "w");
 	fwrite($fd, $ospf6dconffile);
 	fclose($fd);
 
+    //UPGRADE PATH - Checks if legacy "CARP Disable" mode was previously in use proior to 0.6.20; if previously in use, sets "carpmode" to "quaggadisable". If not in use, sets "carpmode" to "none".
+	if (!isset($ospfd_conf['carpmode'])) {
+		syslog(LOG_NOTICE, "Quagga: CARP failover mode upgrade initializing");
+		if (strpos($ospfd_conf['carpstatusvid'],'_vip') !== false) {
+		    $ospfd_conf['carpmode'] = "quaggadisable";
+		    syslog(LOG_NOTICE, "Quagga: CARP failover mode established as 'quaggadisable'");
+		} else {
+		    $ospfd_conf['carpmode'] = "none";
+		    syslog(LOG_NOTICE, "Quagga: CARP failover mode established as 'none'");
+		}
+		write_config( $desc = gettext("Quagga_OSPFd: Upgraded failover method for legacy configurations of Quagga_OSPFD") );
+		syslog(LOG_NOTICE, "Quagga: CARP failover upgrade complete");
+	}
+	
 	$carp_ip_status_check = "";
-	if (isset($ospfd_conf['carpstatusvid']) && $ospfd_conf['carpstatusvid'] != "none") {
+	if ((isset($ospfd_conf['carpmode']) && $ospfd_conf['carpmode'] == "quaggadisable") 
+	    && (isset($ospfd_conf['carpstatusvid']) && $ospfd_conf['carpstatusvid'] != "none")) {
 		$vip = get_configured_vip($ospfd_conf['carpstatusvid']);
 		$carpcheckinterface = escapeshellarg(get_real_interface($vip['interface']));
 		$vhid = escapeshellarg("vhid {$vip['vhid']}");
 		$carp_ip_status_check = <<<EOF
 
-CARP_STATUS=`/sbin/ifconfig {$carpcheckinterface} | /usr/bin/grep 'carp:' | /usr/bin/grep {$vhid} | /usr/bin/awk '{print \$2;}'`
-if [ \${CARP_STATUS} != "MASTER" ]; then
-	exit;
-fi
-EOF;
+		CARP_STATUS=`/sbin/ifconfig {$carpcheckinterface} | /usr/bin/grep 'carp:' | /usr/bin/grep {$vhid} | /usr/bin/awk '{print \$2;}'`
+		if [ \${CARP_STATUS} != "MASTER" ]; then
+			logger "Quagga: CARP \"Quagga Disable\" Mode - Interface {$carpcheckinterface} is NOT in MASTER state, exiting";
+			exit;
+		fi
+		EOF;
```

Comment:
This block should probably not be indented, since it's in a HERE doc. Indenting it will cause the contents to be indented when used, which is probably unintentional.